### DISCUSSION TITLE
Can a filter ignore specific paths ?

### DISCUSSION THREAD

**Discussion by 0x1eef** [nanoc/nanoc#1624] [Get help] [answered]
Hello !

I am working on a filter that runs content through webpack. 
It looks something like this:

```ruby
# Rules
compile "/js/ReactApp.tsx" do
  filter(:webpack, depend_on: ["package-lock.json", "/js/{lib,components,hooks}/**/*.{ts,tsx}"])
  write("/js/app.js")
end
```

And implies that `content/node_modules`, `content/package.json`, and `content/package-lock.json` will exist. I would like to make sure those three paths are always ignored by nanoc. Can a filter do that, or can I only warn about doing it manually in the documentation ? I thought about doing this - assuming that the filter is required in the `Rules` file:

```ruby
ignore "package.json"
ignore "package-lock.json"
ignore "node_modules/**/*"
```

Can I also depend on a file, while ignoring it at the same time?

Thanks!

**Comment by denisdefreyne** [comment, marked as answer]
Hey there,

There is the [`#ignore` convenience rule](https://nanoc.app/doc/rules/#convenience-methods), which makes Nanoc not apply any filters or layouts, and not write it out either. What you wrote it *almost* correct -- it needs a leading slash:

```ruby
ignore "/package.json"
ignore "/package-lock.json"
ignore "/node_modules/**/*"
```

All the items will still be in `@items`. However, Having everything in `@items` might not be ideal, especially with `node_modules` possibly being quite large. You could use a preprocessor rule that removes those items from `@items`:

```ruby
preprocess do
  @items.delete_if { |i| i.identifier.start_with?('/node_modules/') }
end
```

---

Still, my recommendation would be to run asset compilation *outside* of Nanoc, and have Webpack output the resulting `.js`/`.css`/… files into the `content/` directory. Then, you could have a script that invokes Webpack first and then Nanoc.

Webpack (and all the other JS/CSS build tools) are excellent, and I find it easier to run them separately from Nanoc, rather than try to integrate them into Nanoc. (The latter is tough and unreliable, and ultimately I found it not quite worth the effort.)

If you were to use this approach and build your JS/CSS outside of Nanoc and compile into `content/`, you can use the [`#passthrough` convenience method](https://nanoc.app/doc/rules/#convenience-methods):

```ruby
passthrough '/assets/built/**/*'
```

If Webpack were to write the file `content/assets/built/style.6a30406f.css`, then Nanoc would copy that over to `output/assets/built/style.6a30406f.css`. In Nanoc, you’d then be able to link to the stylesheet like this:

```rhtml
  <link rel="stylesheet" href="<%= @items['/assets/built/style.*.css'].path %>">
```

Hope this helps,

Denis

**Comment by 0x1eef** [comment]
Agreed ! In the end I have decided to not store `package.json`, `package-lock.json`, and friends in the `content/` directory. It would be nice to have nanoc rebuild in case a `package.json` dependency changes, but I think there are far more downsides.

> Webpack (and all the other JS/CSS build tools) are excellent, and I find it easier to run them separately from Nanoc, rather than try to integrate them into Nanoc. (The latter is tough and unreliable, and ultimately I found it not quite worth the effort.)

I haven't run into any blockers yet. I am using webpack (through nanoc) in a personal project, and generally it works quite well. It is using TypeScript, and React. The only thing I think I need to improve is investigating why it can take a bit long to compile, sometimes. I think there might be a cache I could use. Not sure.

You can check out the project here: https://github.com/0x1eef/nanoc-webpack.rb - I plan to add sample repositories, eventually. With the `depend_on: ".."` option nanoc will call webpack when some import file changes. Thanks for your help, as always!
